server/target/target.go

```diff
@@ -475,17 +475,13 @@ func fetchTargetsFromOLAPDB(ctx context.Context, env environment.Env, q *query_b
 			},
 			InvocationCreatedAtUsec: row.InvocationStartTimeUsec,
 		})
-		if nextPageToken == nil || nextPageToken.GetInvocationEndTimeUsec() >= row.InvocationStartTimeUsec {
-			nextPageToken = &tppb.PaginationToken{
-				InvocationEndTimeUsec: row.InvocationStartTimeUsec,
-			}
-			if nextPageToken.GetInvocationEndTimeUsec() == row.InvocationStartTimeUsec && nextPageToken.GetCommitSha() > row.CommitSHA {
-				return nil
-			}
-			nextPageToken.CommitSha = row.CommitSHA
+		created_at_usec := commitTimestamps[row.CommitSHA]
+		if row.InvocationStartTimeUsec > created_at_usec {
+			commitTimestamps[row.CommitSHA] = row.InvocationStartTimeUsec
```

Comment:
i'm not quite following--is the core issue that we are merging together all runs for a given commit sha into a single column in the test grid?

Reply:
The way we set the pagination token is wrong; causing some commits won't be selected.

Here is an illustration of the bug:
Let's say we have following data, and the page size is 5. So we are choosing 5 commits per page.
commit1, timestamps: 5000, 4000
commit2, timestamps: 4000, 2000
commit3, timestamps: 3000, 2000
commit4, timestamps: 2000, 1000
commit5, timestamps: 1500, 1000
commit6, timestamps: 1400, 1000
The correct behavior is that the first page will be commit1,commit2, commit3, commit4, commit5; and the next page we will show commit6. The pagination token should be (commit5, 1500)

The current logic will set the pagination token to (commit5, 1000) The following inner select query won't select commit6 because last_created_at_usec of commit6 is 1400 and it's bigger than the timestamp provided in the pagination token (1000). 
```
	//    SELECT commit_sha FROM (
	//      SELECT commit_sha, max(invocation_start_time_usec)
	//      as latest_created_at_usec FROM "TestTargetStatuses"
	//      WHERE group_id = '[group_id]'
	//      AND repo_url = '[repo_url]
	//      AND commit_sha != ''
	//      GROUP BY commit_sha
	//      ORDER BY latest_created_at_usec DESC, commit_sha asc)
	//    WHERE (
	//      latest_created_at_usec < [ts_from_pagination_token] OR
	//      (latest_crated_at_usec = [ts_from_pagination_token] AND
	//      commit_sha > '[commit_sha_from_pagitation_token]'))
	//    LIMIT [page_size]
	```

Reply:
okay, so the core issue is that the old code picked the minimum start time for the invocations in the result set, when it should have been picking `min(max(invocation start time at commit))`? i get it now, thanks for tolerating my questions.